src/protocols/cosmic_workspaces/transactions.c

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <wayland-server-core.h>
+#include <wlr/util/log.h>
+#include "common/list.h"
+#include "common/mem.h"
+#include "protocols/cosmic-workspaces-internal.h"
+
+static void
+transactions_destroy(struct wl_list *list)
+{
+	struct transaction_group *group;
+	struct transaction *trans, *trans_tmp;
+	wl_list_for_each_safe(trans, trans_tmp, list, link) {
+		wlr_log(WLR_INFO, "destroying pending transaction %p", trans);
+		if (trans->changes & CW_PENDING_WS_CREATE) {
+			group = wl_container_of(trans, group, base);
+			free(group->new_workspace_name);
+		}
+		wl_list_remove(&trans->link);
+		free(trans);
+	}
+}
+
+void
+resource_addon_destroy(struct wl_resource_addon *addon)
+{
+	assert(addon);
+	assert(addon->ctx);
+
+	addon->ctx->ref_count--;
+	assert(addon->ctx->ref_count >= 0);
+	wlr_log(WLR_DEBUG, "New refcount for session %p: %d", addon->ctx, addon->ctx->ref_count);
+
+	if (!addon->ctx->ref_count) {
+		wlr_log(WLR_INFO, "Destroying session context");
+		transactions_destroy(&addon->ctx->transactions);
+		free(addon->ctx);
+	}
+
+	free(addon);
+}
+
+struct wl_resource_addon *
+resource_addon_create(struct session_context *ctx)
+{
+	struct wl_resource_addon *addon = znew(*addon);
+	if (!ctx) {
+		ctx = znew(*ctx);
+		wl_list_init(&ctx->transactions);
+	}
+	addon->ctx = ctx;
+	addon->ctx->ref_count++;
+	return addon;
+}
+
+void
+add_ws_transaction(struct wlr_cosmic_workspace *ws,
```

Comment:
Feels like we should start public function names with `transaction_`

Reply:
Those are cosmic internal but can change to transaction prefix.